File: hu/codes/java/chapter_hashing/hash_map_chaining.java
/**
 * File: hash_map_chaining.java
 * Created Time: 2023-06-13
 * Author: krahets (krahets@163.com)
 */

package chapter_hashing;

import java.util.ArrayList;
import java.util.List;

/* Láncolással ütközéskezelő hash tábla */
class HashMapChaining {
    int size; // Kulcs-érték párok száma
    int capacity; // Hash tábla kapacitása
    double loadThres; // Terhelési tényező küszöbértéke a bővítés kiváltásához
    int extendRatio; // Bővítési szorzó
    List<List<Pair>> buckets; // Vödör tömb

    /* Konstruktor */
    public HashMapChaining() {
        size = 0;
        capacity = 4;
        loadThres = 2.0 / 3.0;
        extendRatio = 2;
        buckets = new ArrayList<>(capacity);
        for (int i = 0; i < capacity; i++) {
            buckets.add(new ArrayList<>());
        }
    }

    /* Hash függvény */
    int hashFunc(int key) {
        return key % capacity;
    }

    /* Terhelési tényező */
    double loadFactor() {
        return (double) size / capacity;
    }

    /* Lekérdezési művelet */
    String get(int key) {
        int index = hashFunc(key);
        List<Pair> bucket = buckets.get(index);
        // Vödör bejárása, ha megtalálta a kulcsot, visszaadja a megfelelő értéket
        for (Pair pair : bucket) {
            if (pair.key == key) {
                return pair.val;
            }
        }
        // Ha nem találta meg a kulcsot, null-t ad vissza
        return null;
    }

    /* Hozzáadási művelet */
    void put(int key, String val) {
        // Ha a terhelési tényező meghaladja a küszöbértéket, bővítést hajt végre
        if (loadFactor() > loadThres) {
            extend();
        }
        int index = hashFunc(key);
        List<Pair> bucket = buckets.get(index);
        // Vödör bejárása, ha megtalálta a megadott kulcsot, frissíti a megfelelő értéket és visszatér
        for (Pair pair : bucket) {
            if (pair.key == key) {
                pair.val = val;
                return;
            }
        }
        // Ha a kulcs nem létezik, a kulcs-érték párt a végéhez fűzi
        Pair pair = new Pair(key, val);
        bucket.add(pair);
        size++;
    }

    /* Törlési művelet */
    void remove(int key) {
        int index = hashFunc(key);
        List<Pair> bucket = buckets.get(index);
        // Vödör bejárása és a kulcs-érték pár eltávolítása belőle
        for (Pair pair : bucket) {
            if (pair.key == key) {
                bucket.remove(pair);
                size--;
                break;
            }
        }
    }

    /* Hash tábla bővítése */
    void extend() {
        // Az eredeti hash tábla ideiglenes tárolása
        List<List<Pair>> bucketsTmp = buckets;
        // Bővített új hash tábla inicializálása
        capacity *= extendRatio;
        buckets = new ArrayList<>(capacity);
        for (int i = 0; i < capacity; i++) {
            buckets.add(new ArrayList<>());
        }
        size = 0;
        // Kulcs-érték párok áthelyezése az eredeti hash táblából az újba
        for (List<Pair> bucket : bucketsTmp) {
            for (Pair pair : bucket) {
                put(pair.key, pair.val);
            }
        }
    }

    /* Hash tábla nyomtatása */
    void print() {
        for (List<Pair> bucket : buckets) {
            List<String> res = new ArrayList<>();
            for (Pair pair : bucket) {
                res.add(pair.key + " -> " + pair.val);
            }
            System.out.println(res);
        }
    }
}

public class hash_map_chaining {
    public static void main(String[] args) {
        /* Hash tábla inicializálása */
        HashMapChaining map = new HashMapChaining();

        /* Hozzáadási művelet */
        // Kulcs-érték pár (kulcs, érték) hozzáadása a hash táblához
        map.put(12836, "Xiao Ha");
        map.put(15937, "Xiao Luo");
        map.put(16750, "Xiao Suan");
        map.put(13276, "Xiao Fa");
        map.put(10583, "Xiao Ya");
        System.out.println("\nAfter adding is complete, hash table is\nKey -> Value");
        map.print();

        /* Lekérdezési művelet */
        // Kulcs beírása a hash táblába az érték lekéréséhez
        String name = map.get(13276);
        System.out.println("\nInput student ID 13276, query name " + name);

        /* Törlési művelet */
        // Kulcs-érték pár (kulcs, érték) eltávolítása a hash táblából
        map.remove(12836);
        System.out.println("\nAfter removing 12836, hash table is\nKey -> Value");
        map.print();
    }
}

File: hu/codes/python/chapter_hashing/hash_map_chaining.py
"""
File: hash_map_chaining.py
Created Time: 2023-06-13
Author: krahets (krahets@163.com)
"""

import sys
from pathlib import Path

sys.path.append(str(Path(__file__).parent.parent))
from chapter_hashing.array_hash_map import Pair


class HashMapChaining:
    """Láncolással ütközéskezelő hash tábla"""

    def __init__(self):
        """Konstruktor"""
        self.size = 0  # Kulcs-érték párok száma
        self.capacity = 4  # Hash tábla kapacitása
        self.load_thres = 2.0 / 3.0  # Bővítést kiváltó terhelési tényező küszöbértéke
        self.extend_ratio = 2  # Bővítési szorzó
        self.buckets = [[] for _ in range(self.capacity)]  # Vödör tömb

    def hash_func(self, key: int) -> int:
        """Hash függvény"""
        return key % self.capacity

    def load_factor(self) -> float:
        """Terhelési tényező"""
        return self.size / self.capacity

    def get(self, key: int) -> str | None:
        """Lekérdezési művelet"""
        index = self.hash_func(key)
        bucket = self.buckets[index]
        # A vödörön végigmegy, ha megtalálja a kulcsot, visszaadja a megfelelő értéket
        for pair in bucket:
            if pair.key == key:
                return pair.val
        # Ha a kulcs nem található, None-t ad vissza
        return None

    def put(self, key: int, val: str):
        """Hozzáadási művelet"""
        # Ha a terhelési tényező meghaladja a küszöböt, bővítést végez
        if self.load_factor() > self.load_thres:
            self.extend()
        index = self.hash_func(key)
        bucket = self.buckets[index]
        # A vödörön végigmegy, ha megtalálja a megadott kulcsot, frissíti a megfelelő értéket és visszatér
        for pair in bucket:
            if pair.key == key:
                pair.val = val
                return
        # Ha a kulcs nem létezik, a kulcs-érték párt a végéhez fűzi
        pair = Pair(key, val)
        bucket.append(pair)
        self.size += 1

    def remove(self, key: int):
        """Törlési művelet"""
        index = self.hash_func(key)
        bucket = self.buckets[index]
        # A vödörön végigmegy és eltávolítja a kulcs-érték párt
        for pair in bucket:
            if pair.key == key:
                bucket.remove(pair)
                self.size -= 1
                break

    def extend(self):
        """Hash tábla bővítése"""
        # Az eredeti hash tábla ideiglenes tárolása
        buckets = self.buckets
        # A bővített új hash tábla inicializálása
        self.capacity *= self.extend_ratio
        self.buckets = [[] for _ in range(self.capacity)]
        self.size = 0
        # Kulcs-érték párok áthelyezése az eredeti hash táblából az újba
        for bucket in buckets:
            for pair in bucket:
                self.put(pair.key, pair.val)

    def print(self):
        """Hash tábla kiírása"""
        for bucket in self.buckets:
            res = []
            for pair in bucket:
                res.append(str(pair.key) + " -> " + pair.val)
            print(res)


"""Fő kód"""
if __name__ == "__main__":
    # Hash tábla inicializálása
    hashmap = HashMapChaining()

    # Hozzáadás
    # Kulcs-érték pár (kulcs, érték) hozzáadása a hash táblához
    hashmap.put(12836, "Xiao Ha")
    hashmap.put(15937, "Xiao Luo")
    hashmap.put(16750, "Xiao Suan")
    hashmap.put(13276, "Xiao Fa")
    hashmap.put(10583, "Xiao Ya")
    print("\nHozzáadás után a hash tábla:\n[Kulcs1 -> Érték1, Kulcs2 -> Érték2, ...]")
    hashmap.print()

    # Lekérdezés
    # Kulcs beírása a hash táblába az érték lekérdezéséhez
    name = hashmap.get(13276)
    print("\nA 13276-os diákigazolvány alapján a név: " + name)

    # Törlés
    # Kulcs-érték pár (kulcs, érték) törlése a hash táblából
    hashmap.remove(12836)
    print("\nA 12836-os törlése után a hash tábla:\n[Kulcs1 -> Érték1, Kulcs2 -> Érték2, ...]")
    hashmap.print()
